### PR TITLE
cmd: implement snap run --explain

### PR DESCRIPTION
**Pull request by zyga**
This patch adds a new command line option for "snap run", the new
"--explain" toggle propagates through snap run, snap-confine and
snap-exec. At each stage it shows additional output that explains what
is going on. The elements that are highlighted are of importance to the
execution environment of the application process.

The purpose of this is to educate application developers about,
possibly unfamiliar, things that happen during application startup.

Forum: https://forum.snapcraft.io/t/snap-run-explain/13427
Signed-off-by: Zygmunt Krynicki <me@zygoon.pl>

### CODE REVIEW

**Comment by zyga** [comment]
I need to implement spread test and do a quick review with the people at the sprint.

**Comment by anonymouse64** [comment]
It would be great if the mimics created for layouts were included in `snap run --explain`, it doesn't look like this code changes anything in snap-update-ns to that effect however

**Comment by zyga** [comment]
@anonymouse64 nope, it's just an early prototype.

File: cmd/snap-confine/snap-confine.c [hunk 1]
```diff
@@ -545,6 +562,7 @@ int main(int argc, char **argv)
 	}
 	// Restore process state that was recorded earlier.
 	sc_restore_process_state(&proc_state);
+	sc_explain("Executing %s", invocation.executable);
```
Comment (bboozzoo) [inline cmd/snap-confine/snap-confine.c]: This should probably include some stop marker for the output produced by `explain`.

File: cmd/snap/cmd_run.go [hunk 2]
```diff
@@ -160,6 +178,15 @@ func (x *cmdRun) Execute(args []string) error {
 	if len(args) == 0 {
 		return fmt.Errorf(i18n.G("need the application to run as argument"))
 	}
+	if x.Explain || osutil.GetenvBool("SNAP_EXPLAIN") {
+		enableExplainMode()
+	}
+	explainf("The extra output from snap run --explain is not suitable for parsing.")
```
Comment (bboozzoo) [inline cmd/snap/cmd_run.go]: And this place should include some starting marker.

**Comment by bboozzoo** [comment]
Added some comments about starting and ending markers, otherwise it's not immediately clear where the actual output starts.

```
$ snap run --explain hello-world
The extra output from snap run --explain is not suitable for parsing.

[snap run]

Invoked as "snap"
Inferred application details:
  snap name: hello-world
  app name: hello-world
  confinement: strict
  base: core (implicit)
Will exec through the following tools:
  - snap-confine (execution environment and sandboxing)
  - snap-exec (command chain and environment variables)
Creating user data directories
  - /home/maciek/snap/hello-world/29
  - /home/maciek/snap/hello-world/common
Executing snap-confine /usr/lib/snapd/snap-confine
  with arguments: snap.hello-world.hello-world /usr/lib/snapd/snap-exec hello-world
  with altered environment
    XDG_RUNTIME_DIR: /run/user/1000/snap.hello-world
    SNAP_USER_DATA: /home/maciek/snap/hello-world/29
    SNAP_NAME: hello-world
    SNAP_VERSION: 6.4
    SNAP_USER_COMMON: /home/maciek/snap/hello-world/common
    SNAP_LIBRARY_PATH: /var/lib/snapd/lib/gl:/var/lib/snapd/lib/gl32:/var/lib/snapd/void
    SNAP_ARCH: amd64
    SNAP_REVISION: 29
    SNAP_COMMON: /var/snap/hello-world/common
    SNAP_DATA: /var/snap/hello-world/29
    SNAP: /snap/hello-world/29
    HOME: /home/maciek/snap/hello-world/29
    SNAP_INSTANCE_NAME: hello-world

[snap-confine]

Execution environment:
  - Root file system is /var/lib/snapd/snap/core/current
  - Creating new per-snap mount namespace
    desired fstab /var/lib/snapd/mount/snap.hello-world.fstab
    effective fstab /run/snapd/ns/snap.hello-world.fstab
    info file /run/snapd/ns/snap.hello-world.info
  - Using private /tmp directory
  - Apparmor profile: snap-update-ns.hello-world
    source: /var/lib/snapd/apparmor/profiles/snap-update-ns.hello-world
    binary: (loaded into the kernel, cached by the platform)
  - Apparmor not compiled into snap-confine
  - Creating new per-user mount namespace
    desired user fstab /var/lib/snapd/mount/snap.hello-world.user-fstab
  - Freezer cgroup v1: supported
    path: /sys/fs/cgroup/freezer/hello-world
  - Pids cgroup v1: supported
    path: /sys/fs/cgroup/pids/hello-world
  - Setting PATH to /usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games
  - Setting TMPDIR to /tmp
  - Setting TEMPDIR to /tmp
Sandbox overview
  - Apparmor profile: snap.hello-world.hello-world
    source: /var/lib/snapd/apparmor/profiles/snap.hello-world.hello-world
    binary: (loaded into the kernel, cached by the platform)
  - Apparmor not compiled into snap-confine
  - Applied seccomp profile: snap.hello-world.hello-world
    source: /var/lib/snapd/seccomp/bpf/snap.hello-world.hello-world.src
    binary: /var/lib/snapd/seccomp/bpf/snap.hello-world.hello-world.bin (built with snap-seccomp)
  - Applied seccomp profile: global profile for all snaps
    binary: /var/lib/snapd/seccomp/bpf/global.bin
Executing /usr/lib/snapd/snap-exec
Hello World!
```

File: cmd/snap-confine/snap-confine.c (hunk 1 above)
Comment (zyga) [reply cmd/snap-confine/snap-confine.c]: > mborzecki: thanks for looking at --explain,
> mborzecki: the marker, did you mean to use some kind prefix for each line
> mborzecki: or something else?

Comment (bboozzoo) [reply cmd/snap-confine/snap-confine.c]: As discussed on IRC, I'd try adding something like
```
<< snap explain start >>
...
<< snap explain end >>
... application output follows
```
IIIRC the kernel uses markers like:

```
-------------------- [ cut here ] --------------------
<backtrace>
-------------------- [ cut here ] --------------------
```
The upside is that you don't have to prefix every line, and it's distinct from the application.

**Comment by pedronis** [comment]
@bboozzoo the original sketch (https://forum.snapcraft.io/t/snap-run-explain/13427)  had a marker before the snap app command output:

[snap.hello.hello]

so I think that would have addressed the issue,  maybe we just want to replace the  [] with <<>> for extra visibility across everything.  Also that marker content itself needs some thinking.

There are a lot of punctuation (missing colons especially) and capitalization inconsistencies ATM in this.

File: tests/main/snap-run-explain/expected [hunk 3]
```diff
@@ -0,0 +1,72 @@
+The extra output from snap run --explain is not suitable for parsing.
+
+[snap run]
```
Comment (pedronis) [inline tests/main/snap-run-explain/expected]: as per Maciej suggestion we might consider turning this into:

```
<< snap run >>
```

File: tests/main/snap-run-explain/expected [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+Invoked as "snap"
+Inferred application details:
+  snap name: test-snapd-sh
+  app name: test-snapd-sh
+  confinement: strict
+  base: core (implicit)
+Will exec through the following tools:
+  - snap-confine (execution environment and sandboxing)
+  - snap-exec (command chain and environment variables)
+Creating user data directories
```
Comment (pedronis) [inline tests/main/snap-run-explain/expected]: colon : at the end

File: tests/main/snap-run-explain/expected [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+  - ~/snap/test-snapd-sh/x1
+  - ~/snap/test-snapd-sh/common
+Creating symbolic link "x1" -> "/root/snap/test-snapd-sh/current"
+Executing snap-confine /snap/core/8250/usr/lib/snapd/snap-confine
+  with arguments: snap.test-snapd-sh.test-snapd-sh /usr/lib/snapd/snap-exec test-snapd-sh -c /bin/true
+  with altered environment
```
Comment (pedronis) [inline tests/main/snap-run-explain/expected]: is this the full environment or just what gets added to it ?  if just what's gets added to it  we probably want something different,  also we want a ':' colon here

File: tests/main/snap-run-explain/expected [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    HOME: /root/snap/test-snapd-sh/x1
```
Comment (pedronis) [inline tests/main/snap-run-explain/expected]: Q: was this =  originally and we turned it into :  after comments?

File: tests/main/snap-run-explain/expected [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+    SNAP_INSTANCE_NAME: test-snapd-sh
+    SNAP_LIBRARY_PATH: /var/lib/snapd/lib/gl:/var/lib/snapd/lib/gl32:/var/lib/snapd/void
+    SNAP_NAME: test-snapd-sh
+    SNAP_REVISION: x1
+    SNAP_USER_COMMON: /root/snap/test-snapd-sh/common
+    SNAP_USER_DATA: /root/snap/test-snapd-sh/x1
+    SNAP_VERSION: 1.0
+    XDG_RUNTIME_DIR: /run/user/0/snap.test-snapd-sh
+
+[snap-confine]
```
Comment (pedronis) [inline tests/main/snap-run-explain/expected]: wondering if this should repeat the functionality descriptions:

<< snap-confine (sets up execution environment and sandboxing) >>

File: tests/main/snap-run-explain/expected [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+
+Execution environment:
+  - Root file system is /snap/core/current
+  - Creating new per-snap mount namespace
+    desired fstab /var/lib/snapd/mount/snap.test-snapd-sh.fstab
```
Comment (pedronis) [inline tests/main/snap-run-explain/expected]: "desired fstab: " etc   I find this bit hard to parse as is

File: tests/main/snap-run-explain/expected [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+    effective fstab /run/snapd/ns/snap.test-snapd-sh.fstab
+    info file /run/snapd/ns/snap.test-snapd-sh.info
+  - Using private /tmp directory
+  - Apparmor profile: snap-update-ns.test-snapd-sh
+    source: /var/lib/snapd/apparmor/profiles/snap-update-ns.test-snapd-sh
+    binary: (loaded into the kernel, cached by the platform)
+  - Apparmor not compiled into snap-confine
+
+[snap-update-ns]
```
Comment (pedronis) [inline tests/main/snap-run-explain/expected]: here things get confusing, where does snap-update-ns output finishes ?  maybe

<< snap-update ns (description) >>
<< ... snap-confine >>  once we are back?

File: tests/main/snap-run-explain/expected [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+
+Configuring mount namespace according to mount profile
+  - Freezer cgroup v1: supported
+    path: /sys/fs/cgroup/freezer/test-snapd-sh
+  - Setting PATH to /usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games
+  - Setting TMPDIR to /tmp
+  - Setting TEMPDIR to /tmp
+Sandbox overview
```
Comment (pedronis) [inline tests/main/snap-run-explain/expected]: "Sandbox overview:"   with colon ?

File: tests/main/snap-run-explain/expected [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+  - Apparmor profile: snap.test-snapd-sh.test-snapd-sh
+    source: /var/lib/snapd/apparmor/profiles/snap.test-snapd-sh.test-snapd-sh
```
Comment (pedronis) [inline tests/main/snap-run-explain/expected]: I wonder if a bit more indentation here and similar sublines would help

File: tests/main/snap-run-explain/expected [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+    source: /var/lib/snapd/seccomp/bpf/snap.test-snapd-sh.test-snapd-sh.src
+    binary: /var/lib/snapd/seccomp/bpf/snap.test-snapd-sh.test-snapd-sh.bin (built with snap-seccomp)
+  - Applied seccomp profile: global profile for all snaps
+    binary: /var/lib/snapd/seccomp/bpf/global.bin
+Executing /usr/lib/snapd/snap-exec
+
+[snap exec]
+
+Executing command: /snap/test-snapd-sh/x1/bin/sh
+  with arguments: -c /bin/true
```
Comment (pedronis) [inline tests/main/snap-run-explain/expected]: we want a separator about the starting app here as per original sketch:

<<  test-snapd-sh.sh >>  or [test-snapd-sh.sh]

File: tests/main/snap-run-explain/expected (hunk 9 above)
Comment (pedronis) [reply tests/main/snap-run-explain/expected]: another approach is to consider snap-update-ns really an implementation detail of snap-confine and not prefixing it at all

File: tests/main/snap-run-explain/expected (hunk 5 above)
Comment (zyga) [reply tests/main/snap-run-explain/expected]: Those are just the alterations, which effectively are just the additions (we don't remove environment variables).

File: tests/main/snap-run-explain/expected (hunk 6 above)
Comment (zyga) [reply tests/main/snap-run-explain/expected]: Yes

File: tests/main/snap-run-explain/expected (hunk 8 above)
Comment (zyga) [reply tests/main/snap-run-explain/expected]: Ideas welcome. Everything in the output is dense, the point of this message is that it points (pun not intended) to a file the developer may not have been familiar with before.

File: tests/main/snap-run-explain/expected (hunk 9 above)
Comment (zyga) [reply tests/main/snap-run-explain/expected]: I think it deserves its own section simply because there are sections for snap run / confine / exec (we could reconsider but I think explain was about not hiding it). I agree that the output is a bit intermixed and it needs improved termination.

File: tests/main/snap-run-explain/expected (hunk 11 above)
Comment (zyga) [reply tests/main/snap-run-explain/expected]: I tried to make indentation consistent across all the programs. We can change it consistently across all of them, just need to pick the amount that reads nicely.

File: tests/main/snap-run-explain/expected (hunk 8 above)
Comment (zyga) [reply tests/main/snap-run-explain/expected]: I changed "fstab" to mount profile now.

File: tests/main/snap-run-explain/expected (hunk 4 above)
Comment (zyga) [reply tests/main/snap-run-explain/expected]: Done

File: tests/main/snap-run-explain/expected (hunk 5 above)
Comment (zyga) [reply tests/main/snap-run-explain/expected]: Colon added.

**Comment by zyga** [comment]
I've updated most, looking at the << >> header parts now.

**Comment by zyga** [comment]
@pedronis updated,  there's only a small tweak to match real app name (it's test-snapd-sh, not sh)

File: cmd/snap/cmd_run.go [hunk 13]
```diff
@@ -160,6 +168,13 @@ func (x *cmdRun) Execute(args []string) error {
 	if len(args) == 0 {
 		return fmt.Errorf(i18n.G("need the application to run as argument"))
 	}
+	if x.Explain || osutil.GetenvBool("SNAP_EXPLAIN") {
+		explain.Enable()
+	}
+	explain.Say("The extra output from snap run --explain is not suitable for parsing.")
+	explain.Header("snap run")
+	explain.Say("Invoked as %q", os.Args[0])
```
Comment (pedronis) [inline cmd/snap/cmd_run.go]: this should say Invoked as "snap run",  if os.Args[0] == "snap"   , invoked as "snap" is too terse (too many things are snap in our universe)

File: cmd/explain/explain.go [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+
+// Say prints an explanatory message to standard output.
+//
+// Say is only effective if Enable was called earlier.
+var Say = func(f string, args ...interface{}) {}
+
+// Header prints a spaced header, usually separating subsequent programs.
+//
+// Header  is only effective if Enable was called earlier.
+var Header = func(name string) {}
```
Comment (pedronis) [inline cmd/explain/explain.go]: given the pattern of our headers, this could take name, prefix, summary, to be rendered as:

```<< [$prefix ][$name[ ($comment)] >>```

File: tests/main/snap-run-explain/expected.in [hunk 15]
```diff
@@ -0,0 +1,79 @@
+The extra output from snap run --explain is not suitable for parsing.
```
Comment (niemeyer) [inline tests/main/snap-run-explain/expected.in]: The output below seems to have improved a bit from what was presented in the sprint, but it still feels very inconsistent in terms of:

- Indentation
- Capitalization
- Use of lists vs. not
- Use of parenthesis vs. not
- Use of colon (:) at the end vs. not
- Use of data inline vs. below

In addition:

- The headers seem atypical (<< foo >>)
- Some syntax hints are not entirely readable (e.g. << ... snap-confine >>)
- Better line spacing and indentation >2 would likely improve readability

I think we can do a bit better still.. feel free to book a call if you'd like to go over these points in more detail or brainstorm over something.

**Comment by zyga** [comment]
We just had a call about this PR with @niemeyer and discussed how to approach the structure of the output. Right now the output is crafted mostly manually, by printf-style coding. We need to go through the output that exists today, extract some patterns, like we did with the header API and craft the entire output in terms of those patterns. This will make it easy to have constancy in spacing, capitalisation, interruption and nesting.

There are some specific stylistic tweaks to apply, like having separators with two newlines before, one newline after, deeper indentation for a more pronounced effect, etc but we can adjust that en-mass after the output patterns/templates are implemented.

**Comment by zyga** [comment]
Since the layout code was rejected I'll iterate on this with some bare-bones non-layout, just print and ignore wrapping, code.

**Comment by mvo5** [comment]
I tried a strawman PR with a bit of decoupling of string generation and presentation in #8153 but the review feedback seems to be that we need another approach.

**Comment by pedronis** [comment]
Finding an implementation approach that is not too invasive, produces consistent output and of reasonable complexity has proven quite challenging without being able to make this a main focus for a bit. Closing for now until we have bandwidth to revisit this with the proper attention.
